jcl/src/java.base/share/classes/java/lang/invoke/MethodHandles.java

```diff
@@ -2791,7 +2791,112 @@ public static MethodHandle filterArguments(MethodHandle handle, int startPositio
+	 * The preprocessor handle accepts a subset of the original methods arguments, the subset and their order are 
+	 * dictated by the array <i>argumentIndices</i>.
+	 * 
+	 * Pseudocode example:
+	 * handle: B original(F f, V v, P p)
+	 * filterPosition: 1
+	 * preprocessor:  V filter(V v, F f)
+	 * argumentIndices: {1, 0}
+	 * 
+	 * resulting handle:
+	 * B result(F f, V v, P p) {
+	 * 	V v = filter(V v, F f)
+	 * 	return original(F f, V v, P p)
+	 * }
+	 *
+  	 * @param handle - the handle to call after preprocessing
+ 	 * @param filterPosition - the starting position to filter arguments
+	 * @param preprocessor - a methodhandle that preprocesses some of the incoming arguments
+	 * @param argumentIndices - an array of indices mapping the handle's arguments to the preprocessors inputs
+	 * @return a MethodHandle that preprocesses some of the arguments to the handle
+	 * @throws NullPointerException - if any of the arguments are null
+	 * @throws IllegalArgumentException - if the preprocessor's return type differs from the first argument type of the handle,
+	 *                      or if the arguments taken by the preprocessor isn't a subset of the arguments to the handle
+	 *                      or if the element of argumentIndices is outside of the range of the handle's argument list
+	 *                      or if the arguments specified by argumentIndices from the handle doesn't exactly match the the arguments taken by the preprocessor
+	 */
+	static MethodHandle filterArgumentsWithCombiner(MethodHandle handle, int filterPosition, MethodHandle preprocessor, int... argumentIndices) throws NullPointerException, IllegalArgumentException {
+
+		MethodType handleType = handle.type; // implicit nullcheck
+		MethodType preprocessorType = preprocessor.type; // implicit nullcheck
+		Class<?> preprocessorReturnClass = preprocessorType.returnType;
+		final int handleTypeParamCount = handleType.parameterCount();
+		final int preprocessorTypeParamCount = preprocessorType.parameterCount();
+		final int argIndexCount = argumentIndices.length;
+		
+		if ((filterPosition < 0) || (filterPosition >= handleTypeParamCount)) {
+			/*[MSG "K0637", "The value of {0}: {1} must be in a range from 0 to {2}"]*/
+			throw new IllegalArgumentException(Msg.getString("K0637", new Object[] { //$NON-NLS-1$
+							"the filter position", Integer.toString(filterPosition), //$NON-NLS-1$
+							Integer.toString(handleTypeParamCount)}));
+		}
 	
+		if (preprocessorTypeParamCount != argIndexCount) {
+			/*[MSG "K0638", "The count of argument indices: {0} must be equal to the parameter count of the combiner: {1}"]*/
+			throw new IllegalArgumentException(Msg.getString("K0638", new Object[] { //$NON-NLS-1$
+							Integer.toString(argIndexCount),
+							Integer.toString(preprocessorTypeParamCount)}));
+		}
+		
+		for (int i = 0; i < argIndexCount; i++) {
+			if ((argumentIndices[i] < 0) || (argumentIndices[i] >= handleTypeParamCount)) {
```

Comment:
We need to make a defensive copy of `argumentIndices` or the user could modify the array after we've validated it causing corruption / incorrect behaviour